### ISSUE TITLE
K3s and podman in Kata Container

### ISSUE THREAD

**Issue by Agalin** [kata-containers/kata-containers#8237]
**Is your feature request related to a problem? Please describe.**

Just a few kernel modules are missing in the default kernel to make both podman and K3s (and likely other K8s distributions) usable inside of a kata container (at least a privileged one) without any strange workarounds (like loop device mounts).

Having those two tools supported would make kata usable as our main CI environment - both builds (done through podman) and tests (deployed to in-container k3s clusters) would be usable.

Note: It's about running k3s and podman **inside kata containers** not running kata containers on podman (which has its own issue) or k3s (which is supported).

**Describe the solution you'd like**

Add the following kernel config flags:
* `CONFIG_NETFILTER_XT_MATCH_PHYSDEV` - to make iptables/kubeproxy usable, without it cluster networking doesn't work because kubeproxy cannot create valid rules
* `CONFIG_BRIDGE_NETFILTER` (requirement of the above)
* `CONFIG_KEYS` - for kubelet as it stores crypto keys in the kernel space, this option was enabled in "old style" config files, not sure why it wasn't moved to fragments
* `CONFIG_TMPFS_POSIX_ACL` and `CONFIG_TMPFS_XATTR` - podman requires extended attributes support on the underlying filesystem, it's also true for some other packages (I've seen a related error during a random `apt upgrade` but don't remember which package had troubles), not sure if `POSIX_ACL` is truly necessary but both were always enabled together in the old-style config (similar story to `CONFIG_KEYS` - not moved to fragments)
* `CONFIG_VXLAN` - only if vxlan networking mode must be used, not necessary in my environment

In addition kubelet requires `/dev/kmsg` device to start (and it's not enabled/created in the jail mechanism) but... it doesn't need to be functional. Simple `mknod /dev/kmsg c 1 11` is enough (although OOM kills in the cluster may not be reported as such).

**Describe alternatives you've considered**

* Build custom kernel.
* Rebuild kata-deploy image (or hack the daemonset in a different way) to deploy that custom kernel.

**Additional context**

Tmpfs-related flags and `keys` were both available in kata in the past - before moving to the fragment-based config generation (you can check legacy config templates still available in the repository).

Various other match rules for netfilter are already enabled, I'm not sure why this one isn't (and never was).

**Before raising this feature request**

Have you looked at the [limitations document](https://github.com/kata-containers/kata-containers/blob/main/docs/Limitations.md)?

**Comment by Agalin** [comment]
I'm not completely sure but I believe older podman versions (<4?) didn't require extended attributes on the filesystem.

`CONFIG_NETFILTER_ADVANCED` is [already there](https://github.com/kata-containers/kata-containers/blob/45e82b6581b94031d8eeca94ab6d11913785a5b7/tools/packaging/kernel/configs/fragments/common/netfilter.conf#L7).

**Comment by Agalin** [comment]
We're using native overlayfs so no fuse necessary.

Now that I've checked documentation it seems that XATTR support in tmpfs comes from overlayfs:
> The upper filesystem will normally be writable and if it is it must support the creation of trusted.* and/or user.* extended attributes

**Comment by Agalin** [comment]
Another finding: with properly configured virtiofsd (`--xattr --posix-acl --xattrmap=/map/trusted./user.virtiofs./`) I'm able to run k3s on virtiofs with no need to create a tmpfs mount whatsoever. Unfortunately it doesn't work with podman - something is still not right for it (maybe other permissions missing?) - and I'm failing to manually mount overlayfs.

**Comment by jimlindeman** [comment]
I haven't seen any updates on this for a while... is there a supported "kubernetes for CI" configuration that can run in a Kata privileged container?  The /dev/kmsg issue still exists and breaks minikube kubelet startup (and I assume would break Kind also).

**Comment by jimlindeman** [comment]
FYI, I have found that if we put this logic before starting minikube in a Kata privileged container, then it starts and runs with DockerInDocker (on amd64 nodes):
```
if [ ! -f /dev/fuse ]; then sudo mknod /dev/fuse -m 0666 c 10 229; fi
if [ ! -f /dev/kmsg ]; then sudo mknod /dev/kmsg -m 0666 c 1 11; fi
minikube start -p minikube --driver=docker --bootstrapper=kubeadm --extra-config kubeadm.ignore-preflight-errors=SystemVerification

**Comment by Agalin** [comment]
Supported? No. But we're using kata in CI with both Podman and K3s with the mentioned patches - even switched to dragonball + directvolume-based ephemeral volumes since then - that solved a lot of issues related to using tmpfs as the overlayfs base and IO issues when using virtiofs for overlayfs. Had to switch to static resource allocation due to balloon device related errors but that's not a big deal in our case.

Since introduction of directvolume support we haven't seen any other major issues, there seems to be some kind of cgroups leak but I'd need to check if it still happens on the latest version.
